Server/Services/ILocalTempFileLocks.cs

```diff
@@ -1,28 +1,60 @@
 namespace ThriveDevCenter.Server.Services;
 
 using System;
-using System.Collections.Generic;
 using System.IO;
+using System.Runtime.CompilerServices;
 using System.Threading;
+using System.Threading.Tasks;
+using AsyncKeyedLock;
 using Microsoft.Extensions.Configuration;
 using Microsoft.Extensions.Logging;
 
 public interface ILocalTempFileLocks
 {
     /// <summary>
-    ///   Gets a temporary file path. Note that the receiver should lock the returned string while using the path
-    ///   to avoid multiple places using the same temporary folder at once
+    ///   Gets a temporary file path. Note that the receiver should call <see cref="LockAsync(string)"/>
+    ///   to lock the returned string while using the path so as to avoid multiple places using the same
+    ///   temporary folder at once.
     /// </summary>
     /// <param name="suffix">Suffix to add to the temporary folder</param>
-    /// <param name="path">The final result path</param>
-    /// <returns>A semaphore that needs to be locked while using the path</returns>
-    public SemaphoreSlim GetTempFilePath(string suffix, out string path);
+    /// <returns>The final result path</returns>
+    public string GetTempFilePath(string suffix);
+
+    /// <summary>
+    ///   Asynchronously lock based on a key.
+    /// </summary>
+    /// <param name="path">The file path</param>
+    /// <returns>A disposable value.</returns>
+    public ValueTask<IDisposable> LockAsync(string path);
+
+    /// <summary>
+    ///   Asynchronously lock based on a key, while observing a <see cref="CancellationToken"/>.
+    /// </summary>
+    /// <param name="path">The file path</param>
+    /// <param name="cancellationToken">A cancellation token</param>
+    /// <returns>A disposable value.</returns>
+    public ValueTask<IDisposable> LockAsync(string path, CancellationToken cancellationToken);
+
+    /// <summary>
+    ///   Asynchronously lock based on a key, setting a limit for the <see cref="TimeSpan"/> to wait,
+    ///   while observing a <see cref="CancellationToken"/>.
+    /// </summary>
+    /// <param name="path">The file path</param>
+    /// <param name="timeout">The time limit to wait for</param>
+    /// <param name="cancellationToken">A cancellation token</param>
+    /// <returns>A disposable value.</returns>
+    public ValueTask<AsyncKeyedLockTimeoutReleaser<string>> LockAsync(
+        string path, TimeSpan timeout, CancellationToken cancellationToken);
 }
 
 public class LocalTempFileLocks : ILocalTempFileLocks
 {
     private readonly string baseTempFilePath;
-    private readonly Dictionary<string, SemaphoreSlim> requestedPaths = new();
+    private readonly AsyncKeyedLocker<string> asyncKeyedLocker = new(o =>
+    {
+        o.PoolSize = 20; // sets the max number of pooled semaphores; doesn't affect concurrency
```

Comment:
Just one more thing, could you start this comment with a capital letter and put it on the previous line (trailing comments on code lines is not a preferred style in Thrive)?

Reply:
Done.